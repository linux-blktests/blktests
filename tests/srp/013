#!/bin/bash
#
# Copyright (c) 2018 Western Digital Corporation or its affiliates
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc.

. tests/srp/rc

DESCRIPTION="Direct I/O using a discontiguous buffer"
QUICK=1

discontiguous_io() {
	local byte bytes dev m

	use_blk_mq y y || return $?
	dev=$(get_bdev 0) || return $?
	dd if="${dev}" bs=512 count=1 iflag=direct status=none |
		od -An -v -tu1 -w1 |
		while read -r byte; do
			# shellcheck disable=SC2059
			printf "\\x$(printf "%x" $((byte ^ 0xa5)))"
		done >"${TMPDIR}/data_block"
	wc -c < "${TMPDIR}/data_block"
	src/discontiguous-io -s -w "${dev}" <"${TMPDIR}/data_block" >>"$FULL" ||
		return $?
	dd if="${dev}" bs=512 count=1 iflag=direct status=none |
		cmp - "${TMPDIR}/data_block" ||
		return $?
	src/discontiguous-io -s "${dev}" 2>>"$FULL" |
		cmp - "${TMPDIR}/data_block" ||
		return $?
}

test() {
	trap 'trap "" EXIT; teardown' EXIT
	setup && discontiguous_io && echo Passed
}
