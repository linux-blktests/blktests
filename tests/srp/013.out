Configured SRP target driver
512
Passed
